src/pal/src/sharedmemory/sharedmemory.cpp

```diff
@@ -102,27 +112,49 @@ bool SharedMemoryHelpers::EnsureDirectoryExists(const char *path, bool createIfN
             return false;
         }
 
-        // The path does not exist, create the directory
-        if (mkdir(path, PermissionsMask_AllUsers_ReadWriteExecute) == 0)
+        // The path does not exist, create the directory. The permissions mask passed to mkdir() is filtered by the process'
+        // permissions umask, so mkdir() may not set all of the requested permissions. We need to use chmod() to set the proper
+        // permissions. That creates a race when there is no global lock acquired when creating the directory. Another user's
+        // process may create the directory and this user's process may try to use it before the other process sets the full
+        // permissions. In that case, create a temporary directory first, set the permissions, and rename it to the actual
+        // directory name.
+
+        if (isGlobalLockAcquired)
```

Comment:
The mkdtemp way introduces a new possibility for a file/directory not being deleted from /tmp when the process is killed. For /tmp/.dotnet and /tmp/.dotnet/shm, it's ok because these directories are created only once after /tmp is cleaned up, and they are not deleted thereafter since the shm directory is used for a global lock. For the session directories, they are created when needed and deleted when empty. I'd like to avoid adding more possibilities of stuff not being cleaned up from /tmp, and it's not necessary to create a temp dir and rename when there is a global lock held anyway.